linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IncOptimizer.scala

```diff
@@ -748,17 +837,25 @@ final class IncOptimizer private[optimizer] (config: CommonPhaseConfig, collOps:
         case ApplyStatically(flags, This(), _, _, args) if flags.isConstructor =>
           args.forall(isTriviallySideEffectFree)
 
-        case StoreModule(_, _) => true
-        case _                 => isTriviallySideEffectFree(tree)
+        case StoreModule(_, _) =>
```

Comment:
That's the assumption, yes. I thought `ClassDefChecker` actually checked that we're in the same class. In fact we should even be in the constructor of the same class.

It doesn't prevent inlining because we can never inline the constructor of a module class. The only legit caller of that constructor is the module accessor.

Shall I add the check in `ClassDefChecker`?

Reply:
Huh, our current situation is worse than I thought. According to `ClassDefChecker` and `IRChecker`, `StoreModule` can store any target module with any source value that has the right type. That means in theory we could even have a `StoreModule(..., Null())` ... I have no idea why we designed it that way. At the same time `OptimizerCore` already assumes that `StoreModule(_, _)` is for the current class and with a `This()` rhs, when it tries to eliminate `StoreModule`s in `tryElimStoreModule`.

I'm tempted to simplify the whole thing to have only `StoreModule()` without any argument, with a deserialization hack.

Reply:
Resolved in #4943.